### PR TITLE
Extract uucore::line_ending::LineEnding

### PR DESCRIPTION
**Pull request by simon04**
Aims to provide consistent newline/zero terminator handling.

### CODE REVIEW

File: src/uucore/src/lib/mods/line_ending.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    fn from(line_ending: LineEnding) -> Self {
+        match line_ending {
+            LineEnding::None => panic!("Cannot convert LineEnding::None to u8"),
+            _ => line_ending as Self,
+        }
+    }
+}
+
+impl From<bool> for LineEnding {
+    fn from(is_zero_terminated: bool) -> Self {
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/mods/line_ending.rs:35]: I don't like that this is `from`. A more descriptive name would be nice.

File: src/uu/head/src/head.rs [hunk 2]
```diff
@@ -541,7 +527,7 @@ mod tests {
     #[test]
     fn test_args_modes() {
         let args = options("-n -10M -vz").unwrap();
-        assert!(args.zeroed);
+        assert!(args.line_ending == LineEnding::Nul);
```
Comment (tertsdiepraam) [inline src/uu/head/src/head.rs:530]:
```suggestion
        assert_eq!(args.line_ending, LineEnding::Nul);
```

File: src/uu/head/src/head.rs [hunk 3]
```diff
@@ -561,8 +547,8 @@ mod tests {
         assert!(options("-q").unwrap().quiet);
         assert!(options("--verbose").unwrap().verbose);
         assert!(options("-v").unwrap().verbose);
-        assert!(options("--zero-terminated").unwrap().zeroed);
-        assert!(options("-z").unwrap().zeroed);
+        assert!(options("--zero-terminated").unwrap().line_ending == LineEnding::Nul);
+        assert!(options("-z").unwrap().line_ending == LineEnding::Nul);
```
Comment (tertsdiepraam) [inline src/uu/head/src/head.rs:551]:
```suggestion
        assert_eq!(options("--zero-terminated").unwrap().line_ending, LineEnding::Nul);
        assert_eq!(options("-z").unwrap().line_ending, LineEnding::Nul);
```

File: src/uucore/src/lib/mods/line_ending.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/// Aims to provide consistent newline/zero terminator handling
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/mods/line_ending.rs:1]:
```suggestion
//! Provides consistent newline/zero terminator handling
```

File: src/uucore/src/lib/mods/line_ending.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+use std::fmt::Display;
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+/// Line terminator used for printing and parsing
+pub enum LineEnding {
+    #[default] Newline = b'\n',
+    Nul = 0,
+    Space = b' ',
+    None = 8, // abuse backspace \b to encode None
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/mods/line_ending.rs:11]: This feels like premature optimization to me :)

Besides this does not allow us to have a non-optional line ending in the uutils that need that and it introduces the panic in the `Into<u8>` impl.

File: src/uucore/src/lib/mods/line_ending.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
+/// Aims to provide consistent newline/zero terminator handling
+use std::fmt::Display;
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+/// Line terminator used for printing and parsing
+pub enum LineEnding {
+    #[default] Newline = b'\n',
+    Nul = 0,
+    Space = b' ',
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/mods/line_ending.rs:10]: Is this one used? I don't see it being used in this patch, but maybe I'm missing something.

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Nice idea! I just have some small suggestions.

Just so that the idea is out here (but not necessarily better), we could also make this a newtype around `u8`:
```rust
struct LineEnding(u8);

const NEWLINE = ...;
const ZERO = ...;
```
I like the enum approach too though. Could you also run `cargo fmt`?

File: src/uucore/src/lib/mods/line_ending.rs (hunk 1 above)
Comment (simon04) [reply src/uucore/src/lib/mods/line_ending.rs:34]: I haven't invented this conversion, but simply extracted it from `src/uu/comm/src/comm.rs` 🙂

File: src/uucore/src/lib/mods/line_ending.rs (hunk 6 above)
Comment (simon04) [reply src/uucore/src/lib/mods/line_ending.rs]: Thanks for spotting this. I planed to use it at some point in the past...

File: src/uucore/src/lib/mods/line_ending.rs (hunk 5 above)
Comment (simon04) [reply src/uucore/src/lib/mods/line_ending.rs:11]: I'm not happy with this hack neither. What do you suggest? Using `struct LineEnding(Option(u8));` would be an option?

Comment (tertsdiepraam) [reply src/uucore/src/lib/mods/line_ending.rs:11]: For the utils that need an optional line ending, they could use `Option<LineEnding>` right? Or do all utils need the optional line ending?

File: src/uucore/src/lib/mods/line_ending.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/mods/line_ending.rs:34]: The naming for functions needs to be a bit clearer in `uucore` than in a single util, I suppose. We can keep the method but just rename it. Maybe `from_zero_flag`?

File: src/uucore/src/lib/mods/line_ending.rs (hunk 5 above)
Comment (simon04) [reply src/uucore/src/lib/mods/line_ending.rs:11]: Good idea, thanks! -- Done.

In fact, one once command uses an optional line ending. Thus it was simple to refactor.

File: src/uucore/src/lib/mods/line_ending.rs (hunk 1 above)
Comment (simon04) [reply src/uucore/src/lib/mods/line_ending.rs:34]: Makes sense, yeah. Done.

**Comment by simon04** [comment]
@tertsdiepraam, thank you for taking your time and reviewing my PR. Anything left to be done?

File: src/uu/head/src/head.rs [hunk 7]
```diff
@@ -579,7 +570,7 @@ mod tests {
 
         assert!(!opts.verbose);
         assert!(!opts.quiet);
-        assert!(!opts.zeroed);
+        assert!(opts.line_ending == LineEnding::Newline);
```
Comment (tertsdiepraam) [inline src/uu/head/src/head.rs:573]: Another `assert_eq` here

**Review by tertsdiepraam — COMMENTED** [review]
I like it! Just one last thing as far as I can see (and the CI might complain but we'll see how that goes).

File: src/uu/head/src/head.rs (hunk 7 above)
Comment (simon04) [reply src/uu/head/src/head.rs]: Done. Thank you!

**Review by tertsdiepraam — APPROVED** [review]

**Comment by tertsdiepraam** [comment]
Some small style warnings left, but otherwise ready to be merged!

**Comment by simon04** [comment]
Hi, I'm back from vacation. Thanks for the pointer. I've fixed the remaining clippy warning. Hopefully everything is fine now.

**Comment by tertsdiepraam** [comment]
Nice! Now it's my turn to ask for your feedback. I've pushed a commit with a bit more documentation. Let me know if you agree with what I wrote!

**Comment by simon04** [comment]
Great, thanks for taking care of the documentation. :-)

**Comment by sylvestre** [comment]
bravo for this work :)

**Pull request merged**